### DISCUSSION TITLE
ERC721 setApprovalForAll approves the account.address instead of account.publicKey or user.address

### DISCUSSION THREAD

**Discussion by DWboutin** [alchemyplatform/aa-sdk#813] [Issues] [answered]
Hi! 

We are trying to implement Alchemy's AA in our React.js/Next.js application. We want to use the gas manager to pay for the gas fees of our users.

Right now we are at the proof of concept to know how to make it work correct.

For the login I have this React component managing the login with an hardcoded email. When we click on the button, it sends the email to the email address and when clicking on the email's button it redirects correctly to our website and connects the user.

```typescript
export const HeaderAlchemy = () => {
  const user = useUser()
  const { status } = useSignerStatus()
  const { authenticate } = useAuthenticate()
  const setWalletAddress = useUserStore(state => state.setWalletAddress)

  const isAwaitingEmail = status === "AWAITING_EMAIL_AUTH"

  const login = () => {
    authenticate({ type: "email", email: "email@domain.com" })
  }

  useEffect(() => {
    console.log(user?.address)
    if (user?.address) {
      setWalletAddress(user.address)
    }
  }, [user])

  return (
    <button
      onClick={login}
      className="h-10 rounded-full btn-primary hover:btn-primary-focus px-8 focus:outline-none focus-visible:ring focus-visible:ring-white focus:ring-opacity-20"
      data-test="connect-wallet-button"
      data-testid="header-connect-wallet-btn"
    >
      {!isAwaitingEmail ? "Connect wallet" : "Check email"}
    </button>
  )
}
```

After we are logged in, we try to transfer an NFT from the account, but before we need to approve the the transfer. 

```typescript
  const {
    sendUserOperationAsync,
    sendUserOperationResult,
    isSendingUserOperation,
    error: isSendUserOperationError,
  } = useSendUserOperation({
    client,
    waitForTxn: true,
    onSuccess: ({ hash, request }) => {
      console.log("ON SUCCESS", { hash, request })
      // [optional] Do something with the hash and request
    },
    onError: error => {
      console.log("ON ERROR", { error })
      // [optional] Do something with the error
    },
  })

  const handleNftTransfer = () => {
    const ethersProvider = new ethers.BrowserProvider(window.ethereum)
    const approvalContract = await initContract(
      contractsInfo.voxiesNftItems.address,
      contractsInfo.voxiesNftItems.abi,
      ethersProvider
    )

    const approval = await approvalContract.isApprovedForAll(
      user?.address?.toLowerCase(),
      contractsInfo[contractInfoName].address
    )

    if (approval === false) {
      const approvalData = encodeFunctionData({
        abi: contractsInfo.voxiesNftItems.abi,
        functionName: "setApprovalForAll",
        args: [contractsInfo[contractInfoName].address, true],
      } as any)

      const { hash } = await sendUserOperationAsync({
        uo: {
          target: contractsInfo.voxiesNftItems.address as any,
          data: approvalData,
        },
        account,
      })

      console.log({ hash })
    }
    
    ...rest of the transfer transaction
  }
```

The `setApprovalForAll` transaction launches correctly, but it don't approves the `user.address` but it approves the `account.address` instead. 

![Screenshot_2024-06-27_at_11 48 17_AM](https://github.com/alchemyplatform/aa-sdk/assets/1322006/485320c8-9868-4082-832b-b7467c2db893)

So when we launch a [transferFrom](https://sepolia.etherscan.io/address/0xEfdb94D5F89510FbD3A3d7fB25545922a935e8D1#code#F8#L150) transaction, it gives a `"ERC721: transfer caller is not owner nor approved"` exception.

How can we approve the user address correctly and make this work while using the gas sponsoring feature?

Thank you

**Comment by moldy530** [comment, marked as answer]
ah so the `user?.address?.toLowerCase()` returns the underlying *signer* address. the `account.address` is the address for the NFT itself. 

It's rare that you would actually use the `user.address` and should be using the `account.address` as the account is where data is stored. the `user.address` is the owner of the account, but that wallet is just a traditional EOA and doesn't hold any of the assets

**Reply by DWboutin** [reply]
Damn, I was so close and so far at the same time 😅 

Thank you very much
